File: core/src/com/colorflow/entity/dot/Dot.java
package com.colorflow.entity.dot;

import com.badlogic.gdx.graphics.Texture;
import com.badlogic.gdx.graphics.g2d.Batch;
import com.colorflow.entity.Entity;
import com.colorflow.entity.Path;
import com.colorflow.entity.trigger.Trigger;
import com.colorflow.utility.Position;
import com.colorflow.utility.effect.ExplosionPool;

public class Dot extends Entity {

    private static final Texture stdTexture = new Texture("dots/std.png"),
                                reverseTexture = new Texture("dots/reverse.png"),
                                coinTexture = new Texture("dots/coin.png");
    protected Type type;
    protected Color color;

    public Dot() {
        super();
    }

    public void set(Type type, Color color, Path.Type pathType, Position.Radial start, float speed) {
        this.type = type;
        this.color = color;
        switch (type) {
            case STD:
                this.texture = stdTexture;
                break;
            case REVERSE:
                this.texture = reverseTexture;
                break;
            case COIN:
                this.texture = coinTexture;
                break;
            default:
                throw new IllegalStateException();
        }
        this.initTrail(color);
        this.path.set(pathType, start, speed);
        this.bounds.setRadius(40);
        super.set();
    }

    public Type getType() {
        return type;
    }

    public Color getColour() {
        return color;
    }

    @Override
    public void draw(Batch batch, float parentAlpha) {
        batch.setColor(color.getRGB());
        super.draw(batch, parentAlpha);
        batch.setColor(com.badlogic.gdx.graphics.Color.WHITE);
    }

    @Override
    public void act(float delta) {
        rotateBy(1.5f);
        super.act(delta);
    }

    @Override
    public void destroy(Trigger trigger) {
        ExplosionPool.getInstance().start(getStage(), color.getRGB(), getPosition());
        DotPool.getInstance().free(this);
        super.destroy(trigger);
    }

    @Override
    public void dispose() {
        super.dispose();
    }

    @Override
    public void reset() {
    }

    private void initTrail(Color color) {
        float[] colors = this.trail.getEmitters().first().getTint().getColors();
        colors[0] = color.getRGB().r;
        colors[1] = color.getRGB().g;
        colors[2] = color.getRGB().b;
    }

    public enum Type {
        STD, REVERSE, COIN
    }
}

File: core/src/com/colorflow/play/entity/dot/Dot.kt
package com.colorflow.play.entity.dot

import com.badlogic.gdx.graphics.g2d.Batch
import com.badlogic.gdx.utils.Pool
import com.colorflow.play.entity.Entity
import com.colorflow.play.entity.Path
import com.colorflow.utils.AssetProvider
import com.colorflow.utils.Color
import com.colorflow.utils.Position

class Dot(assets: AssetProvider, pool: Pool<Dot>) : Entity(assets, pool as Pool<Entity>) {
    lateinit var type: Type
        protected set
    lateinit var colour: Color
        protected set

    fun set(type: Type, color: Color, pathType: Path.Type, start: Position.Radial, velocity: Float) {
        this.type = type
        this.colour = color
        when (type) {
            Dot.Type.STD -> this.texture = _assets.get_texture("dot_std")
            Dot.Type.REVERSE -> this.texture = _assets.get_texture("dot_reverse")
            Dot.Type.COIN -> this.texture = _assets.get_texture("dot_coin")
        }
        this.initTrail(color)
        path.type = pathType
        path.pos.x = start.x
        path.pos.y = start.y
        path.velocity = velocity
        this.bounds.setRadius(40f)
        super.set()
    }

    override fun draw(batch: Batch?, parentAlpha: Float) {
        batch!!.color = colour.rgb
        super.draw(batch, parentAlpha)
        batch.color = com.badlogic.gdx.graphics.Color.WHITE
    }

    override fun act(delta: Float) {
        rotateBy(1.5f)
        super.act(delta)
    }

    override fun destroy(cb: (Entity)->Unit) {
        _pool.free(this)
        super.destroy(cb)
    }

    override fun dispose() {
        super.dispose()
    }

    override fun reset() {}

    private fun initTrail(color: Color) {
        val colors = this.trail.emitters.first().tint.colors
        colors[0] = color.rgb.r
        colors[1] = color.rgb.g
        colors[2] = color.rgb.b
    }

    enum class Type {
        STD, REVERSE, COIN
    }
}

